removeOptionalBraces doesn't work in case of partial functions as params

- **Version**: 3.0.0-RC2

- **Configuration**:
```
runner.dialect = "scala3"
rewrite.scala3.convertToNewSyntax = true
rewrite.rules = [RedundantBraces]
rewrite.scala3.removeOptionalBraces = yes
```

## Steps

Given code like this:
```scala
val a = List(Some(1), None, Some(3), Some(4)).map({
  case Some(x) => x match {
    case 1 => "one"
    case _ => "more than one"
  } 
  case _ => "none"
})

```

When I run scalafmt like this:

```bash
scalafmt
```

## Problem

Scalafmt formats code like this:
```scala
org.scalafmt.util.FormatException: java.util.NoSuchElementException
    at org.scalafmt.FormatTests.run(FormatTests.scala:56)
    at org.scalafmt.FormatTests.$anonfun$new$3(FormatTests.scala:33)
    at org.scalafmt.FormatTests.$anonfun$new$3$adapted(FormatTests.scala:33)
    at org.scalafmt.util.CanRunTests.$anonfun$runTest$2(CanRunTests.scala:24)
Caused by: java.util.NoSuchElementException
    at scala.collection.immutable.BitmapIndexedMapNode.apply(HashMap.scala:631)
    at scala.collection.immutable.HashMap.apply(HashMap.scala:132)
    at org.scalafmt.internal.FormatOps.distance(FormatOps.scala:961)
    at org.scalafmt.internal.BestFirstSearch.$anonfun$shouldRecurseOnBlock$1(BestFirstSearch.scala:74)
    at org.scalafmt.internal.BestFirstSearch.$anonfun$shouldRecurseOnBlock$1$adapted(BestFirstSearch.scala:71)
    at scala.Option.filter(Option.scala:319)
    at org.scalafmt.internal.BestFirstSearch.shouldRecurseOnBlock(BestFirstSearch.scala:71)
    at org.scalafmt.internal.BestFirstSearch.shortestPath(BestFirstSearch.scala:158)
    at org.scalafmt.internal.BestFirstSearch.run$1(BestFirstSearch.scala:291)
    at org.scalafmt.internal.BestFirstSearch.getBestPath(BestFirstSearch.scala:292)
    at org.scalafmt.internal.BestFirstSearch$.apply(BestFirstSearch.scala:334)
    at org.scalafmt.Scalafmt$.$anonfun$doFormatOne$5(Scalafmt.scala:128)
    at scala.util.Try$.apply(Try.scala:210)
    at org.scalafmt.Scalafmt$.$anonfun$doFormatOne$3(Scalafmt.scala:128)
    at scala.meta.parsers.Parsed.fold(Errors.scala:12)
    at scala.meta.parsers.Parsed.fold$(Errors.scala:11)
    at scala.meta.parsers.Parsed$Success.fold(Errors.scala:26)
    at org.scalafmt.Scalafmt$.doFormatOne(Scalafmt.scala:124)
    at org.scalafmt.Scalafmt$.doFormat(Scalafmt.scala:102)
    at org.scalafmt.Scalafmt$.formatCode(Scalafmt.scala:71)
    at org.scalafmt.FormatTests.run(FormatTests.scala:44)
```

## Expectation

I would like the formatted output to look like this:
```scala
val a = List(Some(1), None, Some(3), Some(4)).map(
  case Some(x) => x match {
    case 1 => "one"
    case _ => "more than one"
  } 
  case _ => "none"
)
```

## Notes

Not sure how to make it work, I can take a look later on at it, but if @kitbellew you have any suggestions they would help out tremendously.

We did some testing at Scala Center and a bunch of issues came up, some I am trying to fix in scalameta.

--- a/scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala
+++ b/scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala
@@ -2449,7 +2449,8 @@ class FormatOps(
     private def isBlockStart(tree: Term.Block, ft: FormatToken): Boolean =
       tree.stats.headOption.exists(_.tokens.headOption.contains(ft.right))
 
-    @inline private def treeLast(tree: Tree): Option[T] = tree.tokens.lastOption
+    @inline private def treeLast(tree: Tree): Option[T] =
+      tree.tokens.lastOption.map(tokens(_).left)
     @inline private def blockLast(tree: Tree): Option[T] =
       if (isTreeMultiStatBlock(tree)) treeLast(tree) else None
     @inline private def seqLast(seq: Seq[Tree]): Option[T] =

--- a/scalafmt-tests/src/test/resources/scala3/OptionalBraces.stat
+++ b/scalafmt-tests/src/test/resources/scala3/OptionalBraces.stat
@@ -1344,7 +1344,13 @@ val a = List(Some(1), None, Some(3), Some(4)).map({
   case _ => "none"
 })
 >>>
-org.scalafmt.util.FormatException: java.util.NoSuchElementException
+val a = List(Some(1), None, Some(3), Some(4)).map({
+  case Some(x) =>
+    x match
+       case 1 => "one"
+       case _ => "more than one"
+  case _ => "none"
+})
 <<< #2495 2
 rewrite.scala3.removeOptionalBraces = yes
 ===
@@ -1359,4 +1365,12 @@ val bar =
     arr
   })
 >>>
-org.scalafmt.util.FormatException: java.util.NoSuchElementException
+val bar =
+  foo({
+    val arr = new Array[String](10)
+    var x = 0
+    while x < 10 do
+       arr(x) = "x"
+       x += 1
+    arr
+  })

--- a/scalafmt-tests/src/test/resources/scala3/OptionalBraces.stat
+++ b/scalafmt-tests/src/test/resources/scala3/OptionalBraces.stat
@@ -1333,3 +1333,30 @@ object a:
       b &&
       c
    do foo
+<<< #2495 1
+rewrite.scala3.removeOptionalBraces = yes
+===
+val a = List(Some(1), None, Some(3), Some(4)).map({
+  case Some(x) => x match {
+    case 1 => "one"
+    case _ => "more than one"
+  }
+  case _ => "none"
+})
+>>>
+org.scalafmt.util.FormatException: java.util.NoSuchElementException
+<<< #2495 2
+rewrite.scala3.removeOptionalBraces = yes
+===
+val bar =
+  foo({
+    val arr = new Array[String](10)
+    var x = 0
+    while x < 10 do {
+      arr(x) = "x"
+      x += 1
+    }
+    arr
+  })
+>>>
+org.scalafmt.util.FormatException: java.util.NoSuchElementException

--- a/scalafmt-tests/src/test/resources/scala3/OptionalBraces_fold.stat
+++ b/scalafmt-tests/src/test/resources/scala3/OptionalBraces_fold.stat
@@ -1284,3 +1284,30 @@ object a:
    while a &&
       b && c
    do foo
+<<< #2495 1
+rewrite.scala3.removeOptionalBraces = yes
+===
+val a = List(Some(1), None, Some(3), Some(4)).map({
+  case Some(x) => x match {
+    case 1 => "one"
+    case _ => "more than one"
+  }
+  case _ => "none"
+})
+>>>
+org.scalafmt.util.FormatException: java.util.NoSuchElementException
+<<< #2495 2
+rewrite.scala3.removeOptionalBraces = yes
+===
+val bar =
+  foo({
+    val arr = new Array[String](10)
+    var x = 0
+    while x < 10 do {
+      arr(x) = "x"
+      x += 1
+    }
+    arr
+  })
+>>>
+org.scalafmt.util.FormatException: java.util.NoSuchElementException

--- a/scalafmt-tests/src/test/resources/scala3/OptionalBraces_fold.stat
+++ b/scalafmt-tests/src/test/resources/scala3/OptionalBraces_fold.stat
@@ -1295,7 +1295,13 @@ val a = List(Some(1), None, Some(3), Some(4)).map({
   case _ => "none"
 })
 >>>
-org.scalafmt.util.FormatException: java.util.NoSuchElementException
+val a = List(Some(1), None, Some(3), Some(4)).map({
+  case Some(x) =>
+    x match
+       case 1 => "one"
+       case _ => "more than one"
+  case _ => "none"
+})
 <<< #2495 2
 rewrite.scala3.removeOptionalBraces = yes
 ===
@@ -1310,4 +1316,11 @@ val bar =
     arr
   })
 >>>
-org.scalafmt.util.FormatException: java.util.NoSuchElementException
+val bar = foo({
+  val arr = new Array[String](10)
+  var x = 0
+  while x < 10 do
+     arr(x) = "x"
+     x += 1
+  arr
+})

--- a/scalafmt-tests/src/test/resources/scala3/OptionalBraces_keep.stat
+++ b/scalafmt-tests/src/test/resources/scala3/OptionalBraces_keep.stat
@@ -1351,3 +1351,30 @@ object a:
       c
    do
       foo
+<<< #2495 1
+rewrite.scala3.removeOptionalBraces = yes
+===
+val a = List(Some(1), None, Some(3), Some(4)).map({
+  case Some(x) => x match {
+    case 1 => "one"
+    case _ => "more than one"
+  }
+  case _ => "none"
+})
+>>>
+org.scalafmt.util.FormatException: java.util.NoSuchElementException
+<<< #2495 2
+rewrite.scala3.removeOptionalBraces = yes
+===
+val bar =
+  foo({
+    val arr = new Array[String](10)
+    var x = 0
+    while x < 10 do {
+      arr(x) = "x"
+      x += 1
+    }
+    arr
+  })
+>>>
+org.scalafmt.util.FormatException: java.util.NoSuchElementException

--- a/scalafmt-tests/src/test/resources/scala3/OptionalBraces_keep.stat
+++ b/scalafmt-tests/src/test/resources/scala3/OptionalBraces_keep.stat
@@ -1362,7 +1362,13 @@ val a = List(Some(1), None, Some(3), Some(4)).map({
   case _ => "none"
 })
 >>>
-org.scalafmt.util.FormatException: java.util.NoSuchElementException
+val a = List(Some(1), None, Some(3), Some(4)).map({
+  case Some(x) =>
+    x match
+       case 1 => "one"
+       case _ => "more than one"
+  case _ => "none"
+})
 <<< #2495 2
 rewrite.scala3.removeOptionalBraces = yes
 ===
@@ -1377,4 +1383,12 @@ val bar =
     arr
   })
 >>>
-org.scalafmt.util.FormatException: java.util.NoSuchElementException
+val bar =
+  foo({
+    val arr = new Array[String](10)
+    var x = 0
+    while x < 10 do
+       arr(x) = "x"
+       x += 1
+    arr
+  })

--- a/scalafmt-tests/src/test/resources/scala3/OptionalBraces_unfold.stat
+++ b/scalafmt-tests/src/test/resources/scala3/OptionalBraces_unfold.stat
@@ -1474,7 +1474,16 @@ val a = List(Some(1), None, Some(3), Some(4)).map({
   case _ => "none"
 })
 >>>
-org.scalafmt.util.FormatException: java.util.NoSuchElementException
+val a = List(Some(1), None, Some(3), Some(4)).map({
+  case Some(x) =>
+    x match
+       case 1 =>
+         "one"
+       case _ =>
+         "more than one"
+  case _ =>
+    "none"
+})
 <<< #2495 2
 rewrite.scala3.removeOptionalBraces = yes
 ===
@@ -1489,4 +1498,11 @@ val bar =
     arr
   })
 >>>
-org.scalafmt.util.FormatException: java.util.NoSuchElementException
+val bar = foo({
+  val arr = new Array[String](10)
+  var x = 0
+  while x < 10 do
+     arr(x) = "x"
+     x += 1
+  arr
+})

--- a/scalafmt-tests/src/test/resources/scala3/OptionalBraces_unfold.stat
+++ b/scalafmt-tests/src/test/resources/scala3/OptionalBraces_unfold.stat
@@ -1463,3 +1463,30 @@ object a:
       b && c
    do
       foo
+<<< #2495 1
+rewrite.scala3.removeOptionalBraces = yes
+===
+val a = List(Some(1), None, Some(3), Some(4)).map({
+  case Some(x) => x match {
+    case 1 => "one"
+    case _ => "more than one"
+  }
+  case _ => "none"
+})
+>>>
+org.scalafmt.util.FormatException: java.util.NoSuchElementException
+<<< #2495 2
+rewrite.scala3.removeOptionalBraces = yes
+===
+val bar =
+  foo({
+    val arr = new Array[String](10)
+    var x = 0
+    while x < 10 do {
+      arr(x) = "x"
+      x += 1
+    }
+    arr
+  })
+>>>
+org.scalafmt.util.FormatException: java.util.NoSuchElementException

